packages/webdriverio/src/middlewares.js

```diff
@@ -21,31 +22,35 @@ export const elementErrorHandler = (fn) => (commandName, commandFn) => {
                 `that wasn't found, waiting for it...`
             )
 
-            return fn(commandName, () => {
-                /**
-                 * create new promise so we can apply a custom error message in cases waitForExist fails
-                 */
-                return new Promise((resolve, reject) => this.waitForExist().then(resolve, reject)).then(
-                    /**
-                     * if waitForExist was successful requery element and assign elementId to the scope
-                     */
-                    () => {
-                        return this.parent.$(this.selector).then((elem) => {
-                            this.elementId = elem.elementId
-                            return fn(commandName, commandFn).apply(this, args)
-                        })
-                    },
-                    /**
-                     * if waitForExist failes throw custom error
-                     */
-                    () => {
-                        throw new Error(`Can't call ${commandName} on element with selector "${this.selector}" because element wasn't found`)
-                    }
-                )
-            }).apply(this)
+            /**
+             * create new promise so we can apply a custom error message in cases waitForExist fails
+             */
+            try {
+                await this.waitForExist()
+            } catch (error) {
+                throw new Error(
+                    `Can't call ${commandName} on element with selector "${this.selector}" because element wasn't found`)
+            }
+
+            /**
+             * if waitForExist was successful requery element and assign elementId to the scope
+             */
+            const element = await this.parent.$(this.selector);
+            this.elementId = element.elementId
         }
 
-        return fn(commandName, commandFn).apply(this, args)
+        try {
+            return await fn(commandName, commandFn).apply(this, args);
+        } catch(error) {
+            if (error.message.includes("stale element reference")) {
+                const element = await refetchElement(this);
+                this.elementId = element.elementId;
+                this.parent = element.parent;
```

Comment:
you are only updating the `elementId` of the current element and its parent but what happens if you have a higher level chain like `$('foo').$('bar').$('loo')`?

Reply:
I'm blowing away the entire ancestry, and that would be handled by the this.parent = element.parent.